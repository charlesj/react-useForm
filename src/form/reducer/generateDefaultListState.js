import { fromJS, Map } from 'immutable'
import { resolveInitialFieldValue } from './resolveInitialFieldValue'
import { resolveLabel } from './resolveLabel'
import { generateDefaultFieldState } from './generateDefaultFieldState'

/**
 *
 * @param field {import("..").IFormFieldArgs}
 * @param {*} initialValues
 * @param options {import("..").IFormOptions}
 */
export const generateDefaultListState = (field, initialValues, options) => {

  const label = resolveLabel(field, options)
  const value = resolveInitialFieldValue(field.value, initialValues.getIn(field.name.split('.')), field.type)
  const items = getCurrentValues(field.fields, value, options, field.name)

  return fromJS({
    initial: {
      type: field.type,
      optional: field.optional || false,
      label,
    },
    current: {
      helperText: field.helperText || '',
      error: false,
      pristine: true,
      touched: false,
    },
    items,
  }).setIn(['initial', 'field'], field)
    .setIn(['initial', 'options'], options)
}

export const getCurrentValues = (fields = [], initialValues, options, listField) => {
  return initialValues.map((item, index) => {
    return getFields(fields, item, options, listField, index)
  })
}

export const getFields = (fields = [], value = Map(), options, listField, index = 0) => {
  return Map([['fields', fields.reduce((acc, field) => {
    return acc.set(field.name, generateDefaultFieldState(field, value, options))
      .setIn([field.name, 'initial', 'field'], {
        ...field,
        name: `${listField}.items.${index}.fields.${field.name}`,
      })
  }, Map())]])
}
