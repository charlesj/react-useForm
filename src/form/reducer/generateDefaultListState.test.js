import { fromJS } from 'immutable'
import { generateDefaultListState } from './generateDefaultListState'

let field, initialValues, options
beforeEach(() => {
  field = { name: 'fieldName', type: 'list', label: 'List' }
  initialValues = fromJS({})
  options = {}
})

test('simple list', () => {
  expect(generateDefaultListState(field, initialValues, options)).toMatchSnapshot()
})

test('complex list', () => {
  field = {
    name: 'fieldName', type: 'list', label: 'List', fields: [
      { name: 'first', label: 'First' },
      { name: 'second', label: 'Second' },
    ],
  }
  expect(generateDefaultListState(field, initialValues, options)).toMatchSnapshot()
})

test('complex list with initialValues', () => {
  field = {
    name: 'fieldName', type: 'list', label: 'List', fields: [
      { name: 'first', label: 'First' },
      { name: 'second', label: 'Second' },
    ],
  }
  initialValues = fromJS({
    fieldName: [
      { first: '1', second: '2' },
      { first: '10', second: '20' },
    ],
  })
  expect(generateDefaultListState(field, initialValues, options)).toMatchSnapshot()
})
