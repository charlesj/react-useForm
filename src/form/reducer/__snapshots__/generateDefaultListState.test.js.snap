// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`complex list 1`] = `
Immutable.Map {
  "initial": Immutable.Map {
    "type": "list",
    "optional": false,
    "label": "List",
    "field": Object {
      "fields": Array [
        Object {
          "label": "First",
          "name": "first",
        },
        Object {
          "label": "Second",
          "name": "second",
        },
      ],
      "label": "List",
      "name": "fieldName",
      "type": "list",
    },
    "options": Object {},
  },
  "current": Immutable.Map {
    "helperText": "",
    "error": false,
    "pristine": true,
    "touched": false,
  },
  "items": Immutable.List [],
}
`;

exports[`complex list with initialValues 1`] = `
Immutable.Map {
  "initial": Immutable.Map {
    "type": "list",
    "optional": false,
    "label": "List",
    "field": Object {
      "fields": Array [
        Object {
          "label": "First",
          "name": "first",
        },
        Object {
          "label": "Second",
          "name": "second",
        },
      ],
      "label": "List",
      "name": "fieldName",
      "type": "list",
    },
    "options": Object {},
  },
  "current": Immutable.Map {
    "helperText": "",
    "error": false,
    "pristine": true,
    "touched": false,
  },
  "items": Immutable.List [
    Immutable.Map {
      "fields": Immutable.Map {
        "first": Immutable.Map {
          "initial": Immutable.Map {
            "type": "text",
            "value": "1",
            "optional": false,
            "label": "First",
            "field": Object {
              "label": "First",
              "name": "fieldName.items.0.fields.first",
            },
          },
          "current": Immutable.Map {
            "helperText": "",
            "error": false,
            "pristine": true,
            "touched": false,
            "value": "1",
          },
        },
        "second": Immutable.Map {
          "initial": Immutable.Map {
            "type": "text",
            "value": "2",
            "optional": false,
            "label": "Second",
            "field": Object {
              "label": "Second",
              "name": "fieldName.items.0.fields.second",
            },
          },
          "current": Immutable.Map {
            "helperText": "",
            "error": false,
            "pristine": true,
            "touched": false,
            "value": "2",
          },
        },
      },
    },
    Immutable.Map {
      "fields": Immutable.Map {
        "first": Immutable.Map {
          "initial": Immutable.Map {
            "type": "text",
            "value": "10",
            "optional": false,
            "label": "First",
            "field": Object {
              "label": "First",
              "name": "fieldName.items.1.fields.first",
            },
          },
          "current": Immutable.Map {
            "helperText": "",
            "error": false,
            "pristine": true,
            "touched": false,
            "value": "10",
          },
        },
        "second": Immutable.Map {
          "initial": Immutable.Map {
            "type": "text",
            "value": "20",
            "optional": false,
            "label": "Second",
            "field": Object {
              "label": "Second",
              "name": "fieldName.items.1.fields.second",
            },
          },
          "current": Immutable.Map {
            "helperText": "",
            "error": false,
            "pristine": true,
            "touched": false,
            "value": "20",
          },
        },
      },
    },
  ],
}
`;

exports[`simple list 1`] = `
Immutable.Map {
  "initial": Immutable.Map {
    "type": "list",
    "optional": false,
    "label": "List",
    "field": Object {
      "label": "List",
      "name": "fieldName",
      "type": "list",
    },
    "options": Object {},
  },
  "current": Immutable.Map {
    "helperText": "",
    "error": false,
    "pristine": true,
    "touched": false,
  },
  "items": Immutable.List [],
}
`;
